### PR TITLE
Add support for Annotated types

### LINKED ISSUE #217: Binding of Annotated types does not work as one would expect
First of all, my apologies is this issue has already been raised. I've been trying to incorporate the injector framework into one of my projects, but I have hit a snag that I'm not sure how to address. Very succinctly, I have the need to bind the same type to different instances in different parts of the code.

For instance, let's say I have an abstract base class Shape and I have two different places in the code in which I want to inject an implementation:
```
class Shape(ABC):
    pass

@inject
@dataclass
class A:
    first_shape: Shape
    second_shape: Shape  
```    
I want `first_shape` to be an instance of `Square` and `second_shape` to be an instance of `Circle`, both of which inherit from `Shape`. To me, the most natural way to do this would be to annotate these types using `typing.Annotated` and create the respective bindings, like such:

```
@inject
@dataclass
class A:
    first_shape: Annotated[Shape, 'first']
    second_shape: Annotated[Shape, 'second']  
```    
and then create the bindings like this:
```
def configure(self, binder):
    binder.bind(Annotated[Shape, 'first'], to=Circle)
    binder.bind(Annotated[Shape, 'second'], to=Square)
```
This, unfortunately, doesn't work. When I call `injector.get(A)`, I get the following error:
```
injector.CallError: Call to ABCMeta.__new__() failed: Can't instantiate abstract class Shape with abstract method draw (injection stack: ['__main__'])
```

I also tried using  provider methods:
```
    @provider
    def provide_first(self) -> Annotated[Shape, 'first']:
        return Circle()

    @provider
    def provide_second(self) -> Annotated[Shape, 'second']:
        return Square()
```
But this also results in an error:
```
injector.UnknownProvider: couldn't determine provider for typing.Annotated[__main__.Shape, 'first'] to None
```

At this point, my question is whether this is the intended behaviour, and, if so, what is the idiomatic way of binding separate declarations of the same type to different instances?

### PR DESCRIPTION
**Pull request by ljnsn**
Add support for [`Annotated`](https://docs.python.org/3/library/typing.html#typing.Annotated) types.

This is a relatively simple and straightforward implementation with minimal changes and seems to work fine.

Resolves  #217

### CODE REVIEW

File: injector_test.py [hunk 1]
```diff
@@ -1690,3 +1690,30 @@ def function1(a: int | str) -> None:
...
+    class Circle(Shape):
+        pass
+
+    first = Annotated[Shape, "first"]
+
+    def configure(binder):
+        binder.bind(first, to=Circle)
+
+    injector = Injector([configure])
+    assert isinstance(injector.get(first), Circle)
```
Comment (davidparsson) [inline injector_test.py:1719]: I'd like to see another test or two, with multiple bindings to see that Injector treats them as separate binds, for example:
- Making sure that `Annotated[int, "first"]` and `Annotated[int, "second"]` get separate bindings/values
- Making sure that `Annotated[int, "user_id"]` and `int` get separate bindings/values

What do you think?

**Comment by davidparsson** [comment]
There are a few failing tests as well, with different problems for different versions.

File: injector_test.py (hunk 1 above)
Comment (ljnsn) [reply injector_test.py:1719]: Thanks for the feedback! Makes sense, I will add some more tests and take a look at the failing ones.

**Comment by ljnsn** [comment]
@davidparsson have you had time to take a look at the changes? :slightly_smiling_face:

**Comment by ljnsn** [comment]
Sorry about that.. I've now tested with 3.7-3.11 locally and everything passes. There are now less checks to recognise `Annotated` in `_punch_through_alias`, but I think it should be fine (actually, the `isinstance` check is probably sufficient).

The error in `TestThreadSafety` I couldn't reproduce locally.

**Comment by jstasiak** [comment]
That's an interesting feature, I've been wondering about some kind of Annotated support for a while.

I'm wondering if we're not missing some edge cases with this approach that will become obvious later, I'll have another look at this PR when I have a moment

**Comment by ljnsn** [comment]
Have you had time to take a look yet @jstasiak ? I would like to start using this :slightly_smiling_face:

**Comment by tobni** [comment]
This depends on this being true:
```python
assert Annotated[int, "first"] == Annotated[int, "first"]
```
which I didn't know if it were or not, so I checked.

I think this holds due to `_AnnotatedAlias` implementing `__eq__` quite intuitively.

I have checked this to be the case in python 3.10. I have also checked that the impl is order sensitive, e.g
```python
assert Annotated[int, "first", "second"] != Annotated[int, "second", "first"]
```
holds.

A funny corner case due to the equality requirement is that providing `Annotated[int, float("nan")]` is impossible, just as `Annotated[int, object()]` will be.

**Comment by ljnsn** [comment]
@tobni thanks for chiming in on this.

> A funny corner case due to the equality requirement is that providing Annotated[int, float("nan")] is impossible, just as Annotated[int, object()] will be.

Not sure what you mean by this, I just added a test for both of your examples and both seem to work fine.

Note that

```python
x = float("nan")
x == x  # False

y = Annotated[int, float("nan")]
y == y  # True
```

**Comment by tobni** [comment]
Ah I was thinking of 
```python
Annotated[int, object()] == Annotated[int, object()] # False
```
Or similar.

But this is immaterial, sorry for the derail!

**Comment by ljnsn** [comment]
Ok, I see what you mean now, when passing in an annotated type with another instance in the metadata it indeed breaks down. I'd argue that's expected behaviour though, since it's clear that what you're looking up from the injector is not what you bound. It would be better to throw an error in that case though, because currently

```python
foo = Annotated[int, object()]

def configure(binder):
    binder.bind(foo, to=123)

injector = Injector([configure])
injector.get(foo)  # returns 123
injector.get(Annotated[int, object()])  # returns 0
```

Not clear to my why it's returning 0 in the second case, I'll check that.

**Comment by tobni** [comment]
Probably because `Annotated[int, object()]() == 0`, I think that is expected behaviour. But also controversial, see https://github.com/python-injector/injector/issues/116

**Comment by jstasiak** [comment]
Very nice test coverage, I'm convinced this would be good to merge but I need to think of a thing or two first – stay tuned and thanks for your patience.

**Comment by petrkalos** [comment]
+1 to what @ricardo-passthrough mentioned, this PR doesn't work with `@inject`

**Comment by jstasiak** [comment]
The implementation so far has just been released in version 0.22.0.
